framework/quota/src/org/apache/cloudstack/quota/dao/QuotaBalanceDaoImpl.java

```diff
@@ -0,0 +1,199 @@
+//Licensed to the Apache Software Foundation (ASF) under one
+//or more contributor license agreements.  See the NOTICE file
+//distributed with this work for additional information
+//regarding copyright ownership.  The ASF licenses this file
+//to you under the Apache License, Version 2.0 (the
+//"License"); you may not use this file except in compliance
+//with the License.  You may obtain a copy of the License at
+//
+//http://www.apache.org/licenses/LICENSE-2.0
+//
+//Unless required by applicable law or agreed to in writing,
+//software distributed under the License is distributed on an
+//"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+//KIND, either express or implied.  See the License for the
+//specific language governing permissions and limitations
+//under the License.
+package org.apache.cloudstack.quota.dao;
+
+import com.cloud.exception.InvalidParameterValueException;
+import com.cloud.utils.db.Filter;
+import com.cloud.utils.db.GenericDaoBase;
+import com.cloud.utils.db.SearchCriteria;
+import com.cloud.utils.db.Transaction;
+import com.cloud.utils.db.TransactionCallback;
+import com.cloud.utils.db.TransactionLegacy;
+import com.cloud.utils.db.TransactionStatus;
+
+import org.apache.cloudstack.quota.vo.QuotaBalanceVO;
+import org.apache.log4j.Logger;
+import org.springframework.stereotype.Component;
+
+import javax.ejb.Local;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+
+@Component
+@Local(value = { QuotaBalanceDao.class })
+public class QuotaBalanceDaoImpl extends GenericDaoBase<QuotaBalanceVO, Long> implements QuotaBalanceDao {
+    private static final Logger s_logger = Logger.getLogger(QuotaBalanceDaoImpl.class.getName());
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public QuotaBalanceVO findLastBalanceEntry(final Long accountId, final Long domainId, final Date beforeThis) {
+        return Transaction.execute(TransactionLegacy.USAGE_DB, new TransactionCallback<QuotaBalanceVO>() {
+            @Override
+            public QuotaBalanceVO doInTransaction(final TransactionStatus status) {
+                List<QuotaBalanceVO> quotaBalanceEntries = new ArrayList<>();
+                Filter filter = new Filter(QuotaBalanceVO.class, "updatedOn", false, 0L, 1L);
+                SearchCriteria<QuotaBalanceVO> sc = createSearchCriteria();
+                sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+                sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+                sc.addAnd("creditsId", SearchCriteria.Op.EQ, 0);
+                sc.addAnd("updatedOn", SearchCriteria.Op.LT, beforeThis);
+                quotaBalanceEntries = search(sc, filter);
+                return quotaBalanceEntries.size() > 0 ? quotaBalanceEntries.get(0) : null;
```

Comment:
Use ``List.isEmpty()`` rather than checking the ``size`` to determine whether or not a list is empty.  Also, we should check that ``quotaBalanceEntries`` is not ``null`` before checking whether or not it is empty.